src/openmrs-fetch.ts

```diff
@@ -85,6 +86,17 @@ export function openmrsFetch<T = any>(
        * to help developers understand the problem and debug
        */
 
+      /*
+       *Redirect to given url when redirect on auth failure is enabled
+       */
+      const { redirectAuthFailure } = getConfig("@openmrs/esm-api");
+      if (
+        redirectAuthFailure.enabled &&
+        redirectAuthFailure.errors.indexOf(response.status) >= 0
+      ) {
+        window.location.assign(redirectAuthFailure.url);
```

Comment:
Some people will want this to redirect to the refapp login page and others will want it to redirect to the esm-login page within the SPA. Within the SPA, we don't need/want the page to reload when navigating to the login page, since SPA routing can happen without a page reload.

When navigating within an SPA, the appropriate thing to do is call `history.pushState()` instead of `location.assign()`. We need to conditionally call pushState or location.assign() depending on whether the url is in the SPA or not.

The [singleSpa.navigateToUrl()](https://single-spa.js.org/docs/api#navigatetourl) function is a pre-made way of performing that check and then navigating appropriately. Let's use that.

To do so:

```sh
npm install --save single-spa
```

In webpack config, add `'single-spa'` as [an external](https://github.com/openmrs/openmrs-esm-api/blob/master/webpack.config.js#L34).

Reply:
Thanks. I have updated to use `navigateToUrl`.